src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/FunctionDefinitionListener.java

```diff
@@ -0,0 +1,164 @@
+package fi.aalto.cs.apluscourses.intellij.model.task;
+
+import com.intellij.openapi.Disposable;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.editor.EditorFactory;
+import com.intellij.openapi.editor.event.DocumentEvent;
+import com.intellij.openapi.editor.event.DocumentListener;
+import com.intellij.openapi.editor.event.EditorEventMulticaster;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Disposer;
+import com.intellij.openapi.vfs.LocalFileSystem;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import com.intellij.psi.PsiParameter;
+import com.intellij.psi.PsiWhiteSpace;
+import fi.aalto.cs.apluscourses.model.task.ActivitiesListener;
+import fi.aalto.cs.apluscourses.model.task.ListenerCallback;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringTokenizer;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.plugins.scala.lang.psi.api.ScalaPsiElement;
+import org.jetbrains.plugins.scala.lang.psi.api.ScalaRecursiveElementVisitor;
+import org.jetbrains.plugins.scala.lang.psi.api.expr.ScAssignment;
+import org.jetbrains.plugins.scala.lang.psi.api.statements.ScFunctionDefinition;
+import org.jetbrains.plugins.scala.lang.psi.impl.expr.ScReferenceExpressionImpl;
+import org.jetbrains.plugins.scala.lang.psi.impl.statements.params.ScParametersImpl;
+
+
+public class FunctionDefinitionListener implements ActivitiesListener,
```

Comment:
Most of the logic in this class is the same as in `ClassDeclarationListener`. Reuse!